File: analyzers/js-ts/src/__tests__/analyzer.test.ts
import { describe, it, expect, beforeAll, afterAll } from "vitest";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import ts from "typescript";
import { walkFile } from "../ast-walker.js";
import { createDefaultPlugins } from "../framework-plugins/index.js";
import { analyzeRepository } from "../analyzer.js";

function createTempDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "confianalyzer-test-"));
}

function removeTempDir(dir: string): void {
  fs.rmSync(dir, { recursive: true, force: true });
}

function parseAndWalk(code: string, fileName = "test.ts") {
  const sourceFile = ts.createSourceFile(
    fileName,
    code,
    ts.ScriptTarget.ESNext,
    true,
    fileName.endsWith(".tsx") || fileName.endsWith(".jsx") ? ts.ScriptKind.TSX : ts.ScriptKind.TS
  );

  const compilerOptions: ts.CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.ESNext,
    jsx: ts.JsxEmit.React,
    allowJs: true,
  };

  const host = ts.createCompilerHost(compilerOptions);
  const originalGetSourceFile = host.getSourceFile;
  host.getSourceFile = (name, languageVersion) => {
    if (name === fileName) return sourceFile;
    return originalGetSourceFile.call(host, name, languageVersion);
  };

  const program = ts.createProgram([fileName], compilerOptions, host);
  const plugins = createDefaultPlugins();

  return walkFile(sourceFile, program, plugins);
}

describe("AST Walker", () => {
  describe("Basic function extraction", () => {
    it("should extract function declarations", () => {
      const result = parseAndWalk(`
        function greet(name: string): string {
          return "hello " + name;
        }

        export async function fetchData(url: string, retries: number = 3): Promise<any> {
          return fetch(url);
        }

        const add = (a: number, b: number): number => a + b;
      `);

      expect(result.functions).toHaveLength(3);

      const greet = result.functions.find((f) => f.name === "greet");
      expect(greet).toBeDefined();
      expect(greet!.kind).toBe("function");
      expect(greet!.isAsync).toBe(false);
      expect(greet!.isExported).toBe(false);
      expect(greet!.returnType).toBe("string");
      expect(greet!.parameters).toHaveLength(1);
      expect(greet!.parameters[0].name).toBe("name");
      expect(greet!.parameters[0].typeAnnotation).toBe("string");

      const fetchData = result.functions.find((f) => f.name === "fetchData");
      expect(fetchData).toBeDefined();
      expect(fetchData!.isAsync).toBe(true);
      expect(fetchData!.isExported).toBe(true);
      expect(fetchData!.parameters).toHaveLength(2);
      expect(fetchData!.parameters[1].hasDefault).toBe(true);

      const add = result.functions.find((f) => f.name === "add");
      expect(add).toBeDefined();
      expect(add!.parameters).toHaveLength(2);
    });

    it("should extract rest parameters", () => {
      const result = parseAndWalk(`
        function log(...messages: string[]): void {}
      `);

      expect(result.functions[0].parameters[0].isRest).toBe(true);
    });
  });

  describe("Class extraction", () => {
    it("should extract classes with methods", () => {
      const result = parseAndWalk(`
        export abstract class BaseService {
          protected name: string;

          constructor(name: string) {
            this.name = name;
          }

          public async getData(): Promise<string> {
            return this.name;
          }

          private static helper(): void {}
        }

        class UserService extends BaseService implements Serializable {
          getUser(id: number): User {
            return {} as User;
          }
        }
      `);

      expect(result.classes).toHaveLength(2);

      const baseService = result.classes.find((c) => c.name === "BaseService");
      expect(baseService).toBeDefined();
      expect(baseService!.isAbstract).toBe(true);
      expect(baseService!.isExported).toBe(true);
      expect(baseService!.superClass).toBeNull();
      expect(baseService!.methods).toContain("constructor");
      expect(baseService!.methods).toContain("getData");
      expect(baseService!.methods).toContain("helper");

      const userService = result.classes.find((c) => c.name === "UserService");
      expect(userService).toBeDefined();
      expect(userService!.superClass).toBe("BaseService");
      expect(userService!.implements).toContain("Serializable");
      expect(userService!.isExported).toBe(false);

      // Check method FunctionIR details
      const getData = result.functions.find((f) => f.name === "getData");
      expect(getData).toBeDefined();
      expect(getData!.qualifiedName).toBe("BaseService.getData");
      expect(getData!.isAsync).toBe(true);
      expect(getData!.accessibility).toBe("public");

      const helper = result.functions.find((f) => f.name === "helper");
      expect(helper).toBeDefined();
      expect(helper!.isStatic).toBe(true);
      expect(helper!.accessibility).toBe("private");
    });
  });

  describe("Import/Export extraction", () => {
    it("should extract all import types", () => {
      const result = parseAndWalk(`
        import React from "react";
        import { useState, useEffect as effect } from "react";
        import * as path from "path";
        import type { FC } from "react";
      `);

      expect(result.imports.length).toBeGreaterThanOrEqual(3);

      const reactDefault = result.imports.find(
        (i) => i.modulePath === "react" && i.defaultImport === "React"
      );
      expect(reactDefault).toBeDefined();
      expect(reactDefault!.isExternal).toBe(true);

      const reactNamed = result.imports.find(
        (i) => i.modulePath === "react" && i.symbols.length > 0
      );
      expect(reactNamed).toBeDefined();
      const useStateSymbol = reactNamed!.symbols.find((s) => s.name === "useState");
      expect(useStateSymbol).toBeDefined();
      const effectSymbol = reactNamed!.symbols.find((s) => s.name === "useEffect");
      expect(effectSymbol).toBeDefined();
      expect(effectSymbol!.alias).toBe("effect");

      const pathImport = result.imports.find((i) => i.modulePath === "path");
      expect(pathImport).toBeDefined();
      expect(pathImport!.namespaceImport).toBe("path");
    });

    it("should extract require statements", () => {
      const result = parseAndWalk(`
        const express = require("express");
      `);

      expect(result.imports).toHaveLength(1);
      expect(result.imports[0].modulePath).toBe("express");
      expect(result.imports[0].namespaceImport).toBe("express");
      expect(result.imports[0].isExternal).toBe(true);
    });

    it("should extract export declarations", () => {
      const result = parseAndWalk(`
        export function foo() {}
        export const bar = () => {};
        export { foo as baz } from "./other";
        export default function main() {}
      `);

      const fooExport = result.exports.find((e) => e.name === "foo" && !e.fromModule);
      expect(fooExport).toBeDefined();
      expect(fooExport!.isDefault).toBe(false);

      const barExport = result.exports.find((e) => e.name === "bar");
      expect(barExport).toBeDefined();

      const reExport = result.exports.find((e) => e.name === "baz");
      expect(reExport).toBeDefined();
      expect(reExport!.localName).toBe("foo");
      expect(reExport!.fromModule).toBe("./other");

      const defaultExport = result.exports.find((e) => e.name === "main");
      expect(defaultExport).toBeDefined();
      expect(defaultExport!.isDefault).toBe(true);
    });
  });

  describe("Call extraction with stringArgs", () => {
    it("should extract calls with receivers, methods, and string args", () => {
      const result = parseAndWalk(`
        function setup() {
          console.log("hello", "world");
          fetch("/api/users");
          db.query("SELECT * FROM users");
          nested.deep.method("arg1", someVar, "arg2");
        }
      `);

      const logCall = result.calls.find((c) => c.callee === 'console.log');
      expect(logCall).toBeDefined();
      expect(logCall!.receiver).toBe("console");
      expect(logCall!.method).toBe("log");
      expect(logCall!.stringArgs).toEqual(["hello", "world"]);
      expect(logCall!.argumentCount).toBe(2);
      expect(logCall!.enclosingFunction).toBe("setup");

      const fetchCall = result.calls.find((c) => c.callee === "fetch");
      expect(fetchCall).toBeDefined();
      expect(fetchCall!.receiver).toBeNull();
      expect(fetchCall!.method).toBeNull();
      expect(fetchCall!.stringArgs).toEqual(["/api/users"]);

      const queryCall = result.calls.find((c) => c.method === "query");
      expect(queryCall).toBeDefined();
      expect(queryCall!.receiver).toBe("db");
      expect(queryCall!.stringArgs).toEqual(["SELECT * FROM users"]);

      const deepCall = result.calls.find((c) => c.method === "method");
      expect(deepCall).toBeDefined();
      expect(deepCall!.stringArgs).toEqual(["arg1", "arg2"]);
      expect(deepCall!.argumentRefs).toEqual(["someVar"]);
      expect(deepCall!.argumentCount).toBe(3);
    });
  });

  describe("Express route detection", () => {
    it("should detect express routes and set endpointInfo", () => {
      const result = parseAndWalk(`
        function setupRoutes() {
          app.get("/users", getUsers);
          router.post("/users/:id", createUser);
        }
      `);

      // The calls should have express enrichments
      const getCall = result.calls.find(
        (c) => c.method === "get" && c.stringArgs?.includes("/users")
      );
      expect(getCall).toBeDefined();
      expect(getCall!.enrichments).toBeDefined();
      expect(getCall!.enrichments![0].pluginName).toBe("express");
      expect(getCall!.enrichments![0].route).toEqual({ method: "GET", path: "/users" });
      expect(getCall!.enrichments![0].suggestedCategory).toBe("API_ENDPOINT");

      const postCall = result.calls.find(
        (c) => c.method === "post" && c.stringArgs?.includes("/users/:id")
      );
      expect(postCall).toBeDefined();
      expect(postCall!.enrichments![0].route).toEqual({ method: "POST", path: "/users/:id" });

      // The enclosing function should get endpointInfo
      const setupFn = result.functions.find((f) => f.name === "setupRoutes");
      expect(setupFn).toBeDefined();
      expect(setupFn!.endpointInfo).toBeDefined();
      expect(setupFn!.endpointInfo!.method).toBe("GET");
    });
  });

  describe("React component detection", () => {
    it("should detect JSX component renders", () => {
      const result = parseAndWalk(
        `
        function App() {
          return (
            <div>
              <Header title="hello" />
              <UserList />
              <footer />
            </div>
          );
        }
      `,
        "test.tsx"
      );

      const appFn = result.functions.find((f) => f.name === "App");
      expect(appFn).toBeDefined();
      expect(appFn!.enrichments).toBeDefined();
      const reactEnrichment = appFn!.enrichments!.find((e) => e.pluginName === "react");
      expect(reactEnrichment).toBeDefined();
      expect(reactEnrichment!.renders).toContain("Header");
      expect(reactEnrichment!.renders).toContain("UserList");
      // "footer" and "div" are lowercase HTML elements, should NOT be included
      expect(reactEnrichment!.renders).not.toContain("footer");
      expect(reactEnrichment!.renders).not.toContain("div");
      expect(reactEnrichment!.suggestedCategory).toBe("UI_INTERACTION");
    });
  });

  describe("Express false positive avoidance", () => {
    it("should NOT detect apiClient.get as an Express route", () => {
      const result = parseAndWalk(`
        function fetchUsers() {
          apiClient.get("/api/users");
          client.post("/api/orders", data);
          axios.get("/api/items");
          http.delete("/api/items/1");
        }
      `);

      // None of these calls should have an express enrichment
      for (const call of result.calls) {
        const expressEnrichment = call.enrichments?.find((e) => e.pluginName === "express");
        expect(expressEnrichment).toBeUndefined();
      }

      // The function should NOT get endpointInfo from these non-Express calls
      const fn = result.functions.find((f) => f.name === "fetchUsers");
      expect(fn).toBeDefined();
      expect(fn!.endpointInfo).toBeUndefined();
    });
  });

  describe("Axios HTTP call detection", () => {
    it("should detect axios HTTP calls", () => {
      const result = parseAndWalk(`
        function fetchUsers() {
          axios.get("/api/users");
          apiClient.post("/api/orders", data);
        }
      `);

      const getCall = result.calls.find(
        (c) => c.method === "get" && c.stringArgs?.includes("/api/users")
      );
      expect(getCall).toBeDefined();
      const axiosEnrichment = getCall!.enrichments!.find((e) => e.pluginName === "axios");
      expect(axiosEnrichment).toBeDefined();
      expect(axiosEnrichment!.httpCall).toEqual({ method: "GET", urlPattern: "/api/users" });
      expect(axiosEnrichment!.suggestedCategory).toBe("API_CALLER");

      const postCall = result.calls.find(
        (c) => c.method === "post" && c.stringArgs?.includes("/api/orders")
      );
      expect(postCall).toBeDefined();
      const postEnrichment = postCall!.enrichments!.find((e) => e.pluginName === "axios");
      expect(postEnrichment!.httpCall).toEqual({ method: "POST", urlPattern: "/api/orders" });
    });
  });

  describe("Full analyzer integration", () => {
    let tempDir: string;

    beforeAll(() => {
      tempDir = createTempDir();

      // Write test fixture files
      fs.writeFileSync(
        path.join(tempDir, "index.ts"),
        `
        import { Router } from "express";

        export function createApp() {
          const router = Router();
          router.get("/health", (req, res) => {
            res.json({ status: "ok" });
          });
          return router;
        }
        `
      );

      fs.writeFileSync(
        path.join(tempDir, "utils.ts"),
        `
        export const add = (a: number, b: number): number => a + b;
        export const multiply = (a: number, b: number): number => a * b;
        `
      );
    });

    afterAll(() => {
      removeTempDir(tempDir);
    });

    it("should produce a valid IrDocument", async () => {
      const ir = await analyzeRepository(tempDir, "test-repo", false);

      expect(ir.$schema).toBe("confianalyzer-ir-v1");
      expect(ir.version).toBe("0.1.0");
      expect(ir.analyzer.name).toBe("@confianalyzer/js-ts-analyzer");
      expect(ir.analyzer.language).toBe("typescript");
      expect(ir.repository.name).toBe("test-repo");
      expect(ir.files).toHaveLength(2);

      const indexFile = ir.files.find((f) => f.relativePath === "index.ts");
      expect(indexFile).toBeDefined();
      expect(indexFile!.language).toBe("typescript");
      expect(indexFile!.hash).toBeTruthy();
      expect(indexFile!.functions.length).toBeGreaterThan(0);
      expect(indexFile!.imports.length).toBeGreaterThan(0);

      const utilsFile = ir.files.find((f) => f.relativePath === "utils.ts");
      expect(utilsFile).toBeDefined();
      expect(utilsFile!.functions).toHaveLength(2);
      expect(utilsFile!.exports.length).toBeGreaterThanOrEqual(2);
    });
  });

  describe("Incremental analysis", () => {
    let tempDir: string;
    let previousIrPath: string;

    beforeAll(async () => {
      tempDir = createTempDir();

      fs.writeFileSync(
        path.join(tempDir, "stable.ts"),
        `export function stableFunc(): string { return "stable"; }\n`
      );

      fs.writeFileSync(
        path.join(tempDir, "changing.ts"),
        `export function changingFunc(): number { return 1; }\n`
      );

      // Run initial analysis to produce a "previous" IR
      const initialIr = await analyzeRepository(tempDir, "incr-repo", false);
      previousIrPath = path.join(tempDir, "previous-ir.json");
      fs.writeFileSync(previousIrPath, JSON.stringify(initialIr, null, 2), "utf-8");
    });

    afterAll(() => {
      removeTempDir(tempDir);
    });

    it("should reuse unchanged files from previous IR", async () => {
      // Run incremental — no files changed
      const ir = await analyzeRepository(tempDir, "incr-repo", false, previousIrPath);

      expect(ir.files).toHaveLength(2);
      const stableFile = ir.files.find((f) => f.relativePath === "stable.ts");
      expect(stableFile).toBeDefined();
      expect(stableFile!.functions).toHaveLength(1);
      expect(stableFile!.functions[0].name).toBe("stableFunc");
    });

    it("should re-analyze changed files", async () => {
      // Modify one file
      fs.writeFileSync(
        path.join(tempDir, "changing.ts"),
        `export function changingFunc(): number { return 42; }\nexport function newFunc(): void {}\n`
      );

      const ir = await analyzeRepository(tempDir, "incr-repo", false, previousIrPath);

      expect(ir.files).toHaveLength(2);

      // stable.ts should be reused (same hash)
      const stableFile = ir.files.find((f) => f.relativePath === "stable.ts");
      expect(stableFile).toBeDefined();

      // changing.ts should be re-analyzed with new content
      const changingFile = ir.files.find((f) => f.relativePath === "changing.ts");
      expect(changingFile).toBeDefined();
      expect(changingFile!.functions.length).toBe(2);
      expect(changingFile!.functions.find((f) => f.name === "newFunc")).toBeDefined();
    });

    it("should omit deleted files", async () => {
      // Delete one file
      fs.unlinkSync(path.join(tempDir, "changing.ts"));

      const ir = await analyzeRepository(tempDir, "incr-repo", false, previousIrPath);

      // Only stable.ts should remain
      expect(ir.files).toHaveLength(1);
      expect(ir.files[0].relativePath).toBe("stable.ts");

      // Restore the file for other tests
      fs.writeFileSync(
        path.join(tempDir, "changing.ts"),
        `export function changingFunc(): number { return 1; }\n`
      );
    });
  });
});

File: analyzers/go/analyzer_test.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/confianalyzer/analyzer-go/analyzer"
	"github.com/confianalyzer/analyzer-go/irtypes"
)

// helper: create a temp repo with given files and run the analyzer
func analyzeFixture(t *testing.T, files map[string]string) *irtypes.IrDocument {
	t.Helper()
	dir := t.TempDir()

	for name, content := range files {
		path := filepath.Join(dir, name)
		if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
			t.Fatalf("mkdir: %v", err)
		}
		if err := os.WriteFile(path, []byte(content), 0644); err != nil {
			t.Fatalf("write %s: %v", name, err)
		}
	}

	doc, err := analyzer.AnalyzeRepository(dir, "test-repo", false)
	if err != nil {
		t.Fatalf("AnalyzeRepository: %v", err)
	}
	return doc
}

func TestFunctionExtraction(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"main.go": `package main

import "fmt"

func Hello(name string) string {
	return "Hello, " + name
}

func private() {
	fmt.Println("private")
}

type Server struct{}

func (s *Server) Start(port int) error {
	return nil
}
`,
	})

	if len(doc.Files) != 1 {
		t.Fatalf("expected 1 file, got %d", len(doc.Files))
	}

	file := doc.Files[0]
	if len(file.Functions) != 3 {
		t.Fatalf("expected 3 functions, got %d", len(file.Functions))
	}

	// Check Hello function
	hello := file.Functions[0]
	if hello.Name != "Hello" {
		t.Errorf("expected name Hello, got %s", hello.Name)
	}
	if hello.QualifiedName != nil {
		t.Errorf("expected no qualifiedName for top-level func, got %s", *hello.QualifiedName)
	}
	if !hello.IsExported {
		t.Error("expected Hello to be exported")
	}
	if len(hello.Parameters) != 1 {
		t.Fatalf("expected 1 param, got %d", len(hello.Parameters))
	}
	if hello.Parameters[0].Name != "name" {
		t.Errorf("expected param name 'name', got %s", hello.Parameters[0].Name)
	}
	if hello.ReturnType == nil || *hello.ReturnType != "string" {
		t.Errorf("expected return type 'string', got %v", hello.ReturnType)
	}

	// Check private function
	priv := file.Functions[1]
	if priv.Name != "private" {
		t.Errorf("expected name private, got %s", priv.Name)
	}
	if priv.IsExported {
		t.Error("expected private to not be exported")
	}

	// Check method
	start := file.Functions[2]
	if start.Name != "Start" {
		t.Errorf("expected name Start, got %s", start.Name)
	}
	if start.QualifiedName == nil || *start.QualifiedName != "Server.Start" {
		t.Errorf("expected qualifiedName Server.Start, got %v", start.QualifiedName)
	}
}

func TestStructExtraction(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"types.go": `package main

type Base struct {
	ID int
}

type User struct {
	Base
	Name  string
	Email string
}

func (u *User) GetName() string {
	return u.Name
}
`,
	})

	file := doc.Files[0]

	var userClass *irtypes.ClassIR
	for i := range file.Classes {
		if file.Classes[i].Name == "User" {
			userClass = &file.Classes[i]
			break
		}
	}

	if userClass == nil {
		t.Fatal("User class not found")
	}

	if userClass.Kind != "class" {
		t.Errorf("expected kind class, got %s", userClass.Kind)
	}
	if !userClass.IsExported {
		t.Error("expected User to be exported")
	}
	if userClass.SuperClass == nil || *userClass.SuperClass != "Base" {
		t.Errorf("expected superClass Base, got %v", userClass.SuperClass)
	}
	if userClass.IsAbstract {
		t.Error("expected User to not be abstract")
	}
	found := false
	for _, m := range userClass.Methods {
		if m == "GetName" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected GetName in methods, got %v", userClass.Methods)
	}
}

func TestInterfaceExtraction(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"iface.go": `package main

type Reader interface {
	Read(p []byte) (n int, err error)
}

type unexported interface {
	doStuff()
}
`,
	})

	file := doc.Files[0]
	if len(file.Classes) != 2 {
		t.Fatalf("expected 2 classes, got %d", len(file.Classes))
	}

	reader := file.Classes[0]
	if reader.Name != "Reader" {
		t.Errorf("expected name Reader, got %s", reader.Name)
	}
	if !reader.IsAbstract {
		t.Error("expected interface to be abstract")
	}
	if !reader.IsExported {
		t.Error("expected Reader to be exported")
	}
	if len(reader.Methods) != 1 || reader.Methods[0] != "Read" {
		t.Errorf("expected methods [Read], got %v", reader.Methods)
	}

	unexported := file.Classes[1]
	if unexported.IsExported {
		t.Error("expected unexported to not be exported")
	}
}

func TestImportExtraction(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"main.go": `package main

import (
	"fmt"
	"net/http"
	alias "encoding/json"
)

func main() {
	fmt.Println("hello")
	_ = http.StatusOK
	_ = alias.NewDecoder
}
`,
	})

	file := doc.Files[0]
	if len(file.Imports) != 3 {
		t.Fatalf("expected 3 imports, got %d", len(file.Imports))
	}

	fmtImport := file.Imports[0]
	if fmtImport.ModulePath != "fmt" {
		t.Errorf("expected modulePath fmt, got %s", fmtImport.ModulePath)
	}
	if fmtImport.NamespaceImport == nil || *fmtImport.NamespaceImport != "fmt" {
		t.Errorf("expected namespaceImport fmt, got %v", fmtImport.NamespaceImport)
	}
	if !fmtImport.IsExternal {
		t.Error("expected fmt to be external")
	}

	aliasImport := file.Imports[2]
	if aliasImport.ModulePath != "encoding/json" {
		t.Errorf("expected modulePath encoding/json, got %s", aliasImport.ModulePath)
	}
	if aliasImport.NamespaceImport == nil || *aliasImport.NamespaceImport != "alias" {
		t.Errorf("expected namespaceImport alias, got %v", aliasImport.NamespaceImport)
	}
}

func TestExportDetection(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"main.go": `package main

var (
	PublicVar  = "public"
	privateVar = "private"
)

const (
	PublicConst  = 42
	privateConst = 0
)

func ExportedFunc() {}
func unexportedFunc() {}
`,
	})

	file := doc.Files[0]
	exportNames := map[string]bool{}
	for _, exp := range file.Exports {
		exportNames[exp.Name] = true
		if exp.IsDefault {
			t.Errorf("Go should never have default exports, got default for %s", exp.Name)
		}
	}

	if !exportNames["PublicVar"] {
		t.Error("expected PublicVar to be exported")
	}
	if !exportNames["PublicConst"] {
		t.Error("expected PublicConst to be exported")
	}
	if !exportNames["ExportedFunc"] {
		t.Error("expected ExportedFunc to be exported")
	}
	if exportNames["privateVar"] {
		t.Error("expected privateVar to not be exported")
	}
	if exportNames["privateConst"] {
		t.Error("expected privateConst to not be exported")
	}
	if exportNames["unexportedFunc"] {
		t.Error("expected unexportedFunc to not be exported")
	}
}

func TestCallExtraction(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"main.go": `package main

import "fmt"

func doWork() {
	fmt.Println("hello", "world")
	helper(42)
}

func helper(n int) {}
`,
	})

	file := doc.Files[0]
	if len(file.Calls) < 2 {
		t.Fatalf("expected at least 2 calls, got %d", len(file.Calls))
	}

	printlnCall := file.Calls[0]
	if printlnCall.Callee != "fmt.Println" {
		t.Errorf("expected callee fmt.Println, got %s", printlnCall.Callee)
	}
	if printlnCall.Receiver == nil || *printlnCall.Receiver != "fmt" {
		t.Errorf("expected receiver fmt, got %v", printlnCall.Receiver)
	}
	if printlnCall.Method == nil || *printlnCall.Method != "Println" {
		t.Errorf("expected method Println, got %v", printlnCall.Method)
	}
	if len(printlnCall.StringArgs) != 2 || printlnCall.StringArgs[0] != "hello" {
		t.Errorf("expected stringArgs [hello world], got %v", printlnCall.StringArgs)
	}
	if printlnCall.EnclosingFunction == nil || *printlnCall.EnclosingFunction != "doWork" {
		t.Errorf("expected enclosingFunction doWork, got %v", printlnCall.EnclosingFunction)
	}

	helperCall := file.Calls[1]
	if helperCall.Callee != "helper" {
		t.Errorf("expected callee helper, got %s", helperCall.Callee)
	}
	if helperCall.Receiver != nil {
		t.Errorf("expected nil receiver, got %v", helperCall.Receiver)
	}
	if helperCall.ArgumentCount != 1 {
		t.Errorf("expected 1 arg, got %d", helperCall.ArgumentCount)
	}
}

func TestGinRouteDetection(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/test\n\ngo 1.21\n",
		"main.go": `package main

func setupRoutes() {
	r.GET("/api/users", listUsers)
	r.POST("/api/users", createUser)
}

func listUsers() {}
func createUser() {}
`,
	})

	file := doc.Files[0]

	var getCall, postCall *irtypes.CallIR
	for i := range file.Calls {
		if file.Calls[i].Callee == "r.GET" {
			getCall = &file.Calls[i]
		}
		if file.Calls[i].Callee == "r.POST" {
			postCall = &file.Calls[i]
		}
	}

	if getCall == nil {
		t.Fatal("GET call not found")
	}
	if len(getCall.Enrichments) == 0 {
		t.Fatal("expected enrichments on GET call")
	}
	if getCall.Enrichments[0].PluginName != "gin" {
		t.Errorf("expected plugin gin, got %s", getCall.Enrichments[0].PluginName)
	}
	if getCall.Enrichments[0].Route == nil || getCall.Enrichments[0].Route.Method != "GET" {
		t.Error("expected route method GET")
	}
	if getCall.Enrichments[0].Route.Path != "/api/users" {
		t.Errorf("expected route path /api/users, got %s", getCall.Enrichments[0].Route.Path)
	}

	if postCall == nil {
		t.Fatal("POST call not found")
	}
	if len(postCall.Enrichments) == 0 || postCall.Enrichments[0].Route.Method != "POST" {
		t.Error("expected POST enrichment")
	}

	var setupFn *irtypes.FunctionIR
	for i := range file.Functions {
		if file.Functions[i].Name == "setupRoutes" {
			setupFn = &file.Functions[i]
			break
		}
	}
	if setupFn == nil {
		t.Fatal("setupRoutes function not found")
	}
	if setupFn.EndpointInfo == nil {
		t.Error("expected endpointInfo on setupRoutes")
	}
}

func TestFullIRDocument(t *testing.T) {
	doc := analyzeFixture(t, map[string]string{
		"go.mod": "module example.com/myapp\n\ngo 1.21\n",
		"main.go": `package main

import "fmt"

func main() {
	fmt.Println("Hello, world!")
}
`,
	})

	if doc.Schema != "confianalyzer-ir-v1" {
		t.Errorf("expected schema confianalyzer-ir-v1, got %s", doc.Schema)
	}
	if doc.Analyzer.Language != "go" {
		t.Errorf("expected language go, got %s", doc.Analyzer.Language)
	}
	if doc.Repository.Name != "test-repo" {
		t.Errorf("expected repo name test-repo, got %s", doc.Repository.Name)
	}
	if len(doc.Files) != 1 {
		t.Fatalf("expected 1 file, got %d", len(doc.Files))
	}

	// Verify JSON serialization round-trip
	data, err := json.MarshalIndent(doc, "", "  ")
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}

	var parsed irtypes.IrDocument
	if err := json.Unmarshal(data, &parsed); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	if parsed.Schema != doc.Schema {
		t.Error("schema mismatch after round-trip")
	}

	file := doc.Files[0]
	if file.Language != "go" {
		t.Errorf("expected language go, got %s", file.Language)
	}
	if file.Hash == "" {
		t.Error("expected non-empty hash")
	}
	if len(file.Functions) != 1 {
		t.Errorf("expected 1 function, got %d", len(file.Functions))
	}
	if file.Functions[0].Kind != "function" {
		t.Errorf("expected kind function, got %s", file.Functions[0].Kind)
	}
}
